### PR TITLE
Expose entire EndpointBuilder to filter factories

### LINKED ISSUE #43000: Expose entire EndpointBuilder to filter factories
## Background and Motivation

In reaction to #42592, I think we should expose the entire `EndopintBuilder` via `RequestDelegateFactoryOptions`. This would allow downcasting to `RouteEndpointBuilder` and seeing the `RoutePattern` too from filters.

## Proposed API

```diff
namespace Microsoft.AspNetCore.Http;

public sealed class RequestDelegateFactoryOptions
{
-    public IReadOnlyList<Func<EndpointFilterFactoryContext, EndpointFilterDelegate, EndpointFilterDelegate>>? EndpointFilterFactories { get; init; }
-    public IList<object>? EndpointMetadata { get; init; }
+    public EndpointBuilder? EndpointBuilder { get; init; }
}
```

```diff
namespace Microsoft.AspNetCore.Http;

public sealed class EndpointFilterFactoryContext
{
-    public EndpointFilterFactoryContext(MethodInfo methodInfo, IList<object> endpointMetadata, IServiceProvider applicationServices);
+    public EndpointFilterFactoryContext(MethodInfo methodInfo, EndpointBuilder endopintBuilder);

-    public IList<object> EndpointMetadata { get; }
-    public IServiceProvider ApplicationServices { get; }
+    public EndpointBuilder EndpointBuilder { get; }
}
```

## Usage Examples

Before:

```C#
RequestDelegateFactoryOptions factoryOptions = new()
{
    ServiceProvider = _applicationServices,
    RouteParameterNames = routeParamNames,
    ThrowOnBadRequest = _throwOnBadRequest,
    DisableInferBodyFromParameters = ShouldDisableInferredBodyParameters(entry.HttpMethods),
    EndpointMetadata = builder.Metadata,
    // builder.EndopintFilterFactories is not settable so is always a List.
    EndpointFilterFactories = (IReadOnlyList<Func<EndpointFilterFactoryContext, EndpointFilterDelegate, EndpointFilterDelegate>>)builder.EndpointFilterFactories,
};
```

After:

```C#
RequestDelegateFactoryOptions factoryOptions = new()
{
    ServiceProvider = _applicationServices,
    RouteParameterNames = routeParamNames,
    ThrowOnBadRequest = _throwOnBadRequest,
    DisableInferBodyFromParameters = ShouldDisableInferredBodyParameters(entry.HttpMethods),
    EndpointBuilder = builder,
};
```

## Alternative Designs

We could keep the filter factories directly on `RequestDelegateFactoryOptions` but rename `EndpointFilterFactories ` to `FilterFactories` to match the name in `EndpointBuilder` and make it an `IList` instead of `IReadOnlyList` which aligns with `RequestDelegateFactoryOptions.Metadata`.

```diff
namespace Microsoft.AspNetCore.Http;

public sealed class RequestDelegateFactoryOptions
{
-    public IReadOnlyList<Func<EndpointFilterFactoryContext, EndpointFilterDelegate, EndpointFilterDelegate>>? EndpointFilterFactories { get; init; }
+    public IList<Func<EndpointFilterFactoryContext, EndpointFilterDelegate, EndpointFilterDelegate>>? FilterFactories { get; init; }
}
```

We could also keep it as an `IReadOnlyList`.  Arguably this communicates better that the `RequestDelegateFactory` won't mutate it. RDF *does* mutate `EndpointMetadata` which is an `IList` so that might add to the confusion.

## Risks

Low. This affects API that has only been released in previews and few people depend on.

### PR DESCRIPTION
**Pull request by halter73**
This flows the entire `EndpointBuilder` through to endpoint filter factories. This gives more context like the display name to filters, and reduces the number of properties in `RequestDelegateFactoryOptions` and `EndpointFilterFactoryContext`. And since the `EndpointBuilder` is usually a `RouteEndpointBuilder`, filters can observe the route pattern as well.

`EndpointBuilder.RequestDelegate` must be `null` if it is passed into the `RequestDelegateFactory.Create()`. The `RequestDelegateFactory` sets the `RequestDelegate` to the same instance returned via `RequestDelegateResult.RequestDelegate`. This is similar to how we were already mutating the passed-in metadata that was also then returned.

Fixes #43000

### CODE REVIEW

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 1]
```diff
@@ -248,15 +273,15 @@ private static IReadOnlyList<object> AsReadOnlyList(IList<object> metadata)
...
+            factoryContext.EndpointBuilder.Metadata,
             factoryContext.ServiceProvider,
             CollectionsMarshal.AsSpan(factoryContext.Parameters));
 
         EndpointFilterDelegate? filterPipeline = null;
 
         // If there are filters registered on the route handler, then we update the method call and
         // return type associated with the request to allow for the filter invocation pipeline.
-        if (factoryContext.FilterFactories is { Count: > 0 })
+        if (factoryContext.EndpointBuilder.FilterFactories.Count > 0 )
```
Comment (halter73) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:284]:
```suggestion
        if (factoryContext.EndpointBuilder.FilterFactories.Count > 0)
```

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 2]
```diff
@@ -842,7 +864,7 @@ private static Expression CreateParamCheckingResponseWritingMethodCall(Type retu
 
         // If filters have been registered, we set the `wasParamCheckFailure` property
         // but do not return from the invocation to allow the filters to run.
-        if (factoryContext.FilterFactories is { Count: > 0 })
+        if (factoryContext.EndpointBuilder.FilterFactories.Count > 0 )
```
Comment (halter73) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:867]:
```suggestion
        if (factoryContext.EndpointBuilder.FilterFactories.Count > 0)
```

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 3]
```diff
@@ -173,46 +183,61 @@ public static RequestDelegateResult Create(MethodInfo methodInfo, Func<HttpConte
...
 
-        return new RequestDelegateResult(httpContext => targetableRequestDelegate(targetFactory(httpContext), httpContext), AsReadOnlyList(factoryContext.Metadata));
+        return CreateRequestDelegateResult(finalRequestDelegate, factoryContext.EndpointBuilder);
     }
 
     private static FactoryContext CreateFactoryContext(RequestDelegateFactoryOptions? options, Delegate? handler = null)
     {
+        if (options?.EndpointBuilder?.RequestDelegate is not null)
+        {
+            throw new ArgumentException($"{nameof(RequestDelegateFactoryOptions)}.{nameof(RequestDelegateFactoryOptions.EndpointBuilder)} must be null.", nameof(options));
```
Comment (BrennanConroy) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:217]: Did you mean to include `RequestDelegate` in this exception message?

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 4]
```diff
@@ -138,17 +139,26 @@ public static RequestDelegateResult Create(Delegate handler, RequestDelegateFact
...
+        RequestDelegate finalRequestDelegate = targetableRequestDelegate switch
         {
             // handler is a RequestDelegate that has not been modified by a filter. Short-circuit and return the original RequestDelegate back.
             // It's possible a filter factory has still modified the endpoint metadata though.
-            return new RequestDelegateResult((RequestDelegate)handler, AsReadOnlyList(factoryContext.Metadata));
+            null => (RequestDelegate)handler,
+            _ => httpContext => targetableRequestDelegate(handler.Target, httpContext),
+        };
+
+        if (targetableRequestDelegate is null)
```
Comment (BrennanConroy) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:152]: 😄 couldn't decide whether to use a switch expression or if statement so went with both?

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 5]
```diff
@@ -173,46 +183,61 @@ public static RequestDelegateResult Create(MethodInfo methodInfo, Func<HttpConte
...
     }
 
     private static FactoryContext CreateFactoryContext(RequestDelegateFactoryOptions? options, Delegate? handler = null)
     {
+        if (options?.EndpointBuilder?.RequestDelegate is not null)
+        {
+            throw new ArgumentException($"{nameof(RequestDelegateFactoryOptions)}.{nameof(RequestDelegateFactoryOptions.EndpointBuilder)} must be null.", nameof(options));
+        }
+
+        var endointBuilder = options?.EndpointBuilder ?? new RDFEndpointBuilder();
```
Comment (BrennanConroy) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:220]:
```suggestion
        var endpointBuilder = options?.EndpointBuilder ?? new RDFEndpointBuilder();
```

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 6]
```diff
@@ -173,46 +183,61 @@ public static RequestDelegateResult Create(MethodInfo methodInfo, Func<HttpConte
...
+        var serviceProvider = options?.ServiceProvider ?? endointBuilder.ApplicationServices;
```
Comment (BrennanConroy) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:221]:
```suggestion
        var serviceProvider = options?.ServiceProvider ?? endpointBuilder.ApplicationServices;
```

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 7]
```diff
@@ -173,46 +183,61 @@ public static RequestDelegateResult Create(MethodInfo methodInfo, Func<HttpConte
...
-            ServiceProvider = options?.ServiceProvider,
-            ServiceProviderIsService = options?.ServiceProvider?.GetService<IServiceProviderIsService>(),
+            ServiceProvider = serviceProvider,
+            ServiceProviderIsService = serviceProvider.GetService<IServiceProviderIsService>(),
             RouteParameters = options?.RouteParameterNames?.ToList(),
             ThrowOnBadRequest = options?.ThrowOnBadRequest ?? false,
             DisableInferredFromBody = options?.DisableInferBodyFromParameters ?? false,
-            FilterFactories = options?.EndpointFilterFactories?.ToList(),
-            Metadata = options?.EndpointMetadata ?? new List<object>(),
+            EndpointBuilder = endointBuilder,
```
Comment (BrennanConroy) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:231]:
```suggestion
            EndpointBuilder = endpointBuilder,
```

**Comment by halter73** [comment]
Closing in favor of #43543.

**Pull request closed without merging**